processor/src/decorators/adv_stack_injectors.rs

```diff
@@ -558,6 +567,190 @@ where
+                        new_root
+                    }
+                };
+
+                self.advice_provider.push_stack(AdviceSource::Word(leaf_value))?;
+                self.advice_provider.push_stack(AdviceSource::Word(new_root))?;
+
+                self.advice_provider.push_stack(AdviceSource::Value(ZERO))?;
+                self.advice_provider.push_stack(AdviceSource::Value(ZERO))?;
+
+                self.advice_provider.push_stack(AdviceSource::Value(ZERO))?;
+                self.advice_provider.push_stack(AdviceSource::Value(ZERO))?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Returns info about a lone sibling of a leaf specified by depth and index parameters in the
+    /// Tiered Sparse Merkle tree defined by the specified root. If no lone siblings exist for the
+    /// specified parameters, None is returned.
+    ///
+    /// A lone sibling is defined as a leaf which has a common root with the specified leaf at a
+    /// higher tier such that the subtree starting at this root contains only these two leaves.
+    ///
+    /// In addition to the leaf node itself, this also returns the index of the common root at a
+    /// higher tier.
+    fn find_lone_sibling(
+        &self,
+        root: Word,
+        depth: u8,
+        index: &Felt,
+    ) -> Result<Option<(RpoDigest, NodeIndex)>, ExecutionError> {
+        debug_assert!(matches!(depth, 16 | 32 | 48));
+
+        // if the leaf is on the first tier (depth=16), we don't care about lone siblings as they
+        // cannot be moved to a higher tier.
+        if depth == TieredSmt::TIER_SIZE {
+            return Ok(None);
+        }
+
+        let empty = &EmptySubtreeRoots::empty_hashes(TieredSmt::MAX_DEPTH)[..=depth as usize];
+
+        // get the path to the leaf node
+        let path: Vec<_> = self.advice_provider.get_merkle_path(root, &depth.into(), index)?.into();
+
+        // traverse the path from the leaf up to the root, keeping track of all non-empty nodes;
+        // here we ignore the top 16 depths because lone siblings cannot be moved to a higher tier
+        // from tier at depth 16.
+        let mut non_empty_nodes = BTreeMap::new();
+        for (depth, p_node) in (TieredSmt::TIER_SIZE..=depth).rev().zip(path.iter()) {
+            // map the depth of each node to the tier it would "round up" to. For example, 17 maps
+            // to tier 1, 32 also maps to tier 1, but 33 maps to tier 2.
+            let tier = (depth - 1) / TieredSmt::TIER_SIZE;
+
+            // if the node is non-empty insert it into the map, but if a node for the same tier is
+            // already in the map, return. We can do this because if a path has two non-empty nodes
+            // at a given tier, then the node is not a lone sibling.
+            let e_node = &empty[depth as usize];
+            if p_node != e_node && non_empty_nodes.insert(tier, (depth, *p_node)).is_some() {
+                return Ok(None);
+            }
```

Comment:
I'm trying to reason about this condition:

```rust
            if p_node != e_node && non_empty_nodes.insert(tier, (depth, *p_node)).is_some() {
                return Ok(None);
            }
```

If we have a case where we have four nodes that all share the same 30 bit prefix.  Two of them also share a 40 bit prefix. As such we have two nodes at tier 3 and two nodes at tier 2.  I'm trying to consider the case where we remove one of the nodes at tier 3.  On the path from the removed leaf to depth 16 we encounter both non empty nodes associated with the leaves at tier 2.  Wouldn't this trigger the the short circuit for  `non_empty_nodes.insert(tier, (depth, *p_node)).is_some()`? If this is the case then I think we may be incorrectly reporting that there is no lone sibling.  Maybe this condition should only be relevant for the lowest tier?

Reply:
Supporting test here: https://github.com/0xPolygonMiden/miden-vm/commit/41dc9312d2de9707be28d40928dc3cc6da198039

Reply:
This is a really good catch! Fixed!